app/src/shared/utils/withWeb3.jsx

```diff
@@ -0,0 +1,102 @@
+// @flow
+
+import React, { type ComponentType } from 'react'
+
+import getWeb3, { validateWeb3 } from '$shared/web3/web3Provider'
+import Web3ErrorDialog from '$shared/components/Web3ErrorDialog'
+import { WalletLockedError } from '$shared/errors/Web3'
+
+import Web3Poller from '$shared/web3/web3Poller'
+import type { Props as DialogProps } from '$shared/components/Dialog'
+
+type State = {
+    checkingWeb3: boolean,
+    web3Error: ?Error,
+}
+
+type OwnProps = {
+    requireWeb3: boolean,
+    onCancel?: () => void,
+}
+
+type Props = DialogProps & OwnProps
+
+export function withWeb3(WrappedComponent: ComponentType<any>) {
+    class WithWeb3 extends React.Component<Props, State> {
+        static defaultProps = {
+            requireWeb3: true,
+        }
+
+        state = {
+            checkingWeb3: false,
```

Comment:
perhaps:
```js
checkingWeb3: this.props.requireWeb3
```
This skips the first render with `checkingWeb3: false` immediately re-rendering with `checkingWeb3: true`

Reply:
It needs to be `false` to start the initial validation but I could add a flag to the `validateWeb3` method.